imodels/tree/hierarchical_shrinkage.py

```diff
@@ -225,14 +226,36 @@ def __init__(self, estimator_: BaseEstimator = None,
         #     raise Warning('Passed an already fitted estimator,'
         #                   'but shrinking not applied until fit method is called.')
 
-    def fit(self, X, y, *args, **kwargs):
+    def fit_slow(self, X, y, *args, **kwargs):
+        s = time.time()
         self.scores_ = []
         for reg_param in self.reg_param_list:
             est = HSTreeClassifier(deepcopy(self.estimator_), reg_param)
             cv_scores = cross_val_score(est, X, y, cv=self.cv, scoring=self.scoring)
             self.scores_.append(np.mean(cv_scores))
         self.reg_param = self.reg_param_list[np.argmax(self.scores_)]
         super().fit(X=X, y=y, *args, **kwargs)
+        e = time.time() - s
+        print(f"slow time: {np.round(e, 2)}")
+
+    def fit(self, X, y, *args, **kwargs):
+        s = time.time()
+        self.scores_ = [[] for _ in self.reg_param_list]
+        kf = KFold(n_splits=self.cv)
+        for train_index, test_index in kf.split(X):
+            X_out, y_out = X[test_index, :], y[test_index]
+            X_in, y_in = X[train_index, :], y[train_index]
+            base_est = deepcopy(self.estimator_)
+            base_est.fit(X_in, y_in)
+            for i, reg_param in enumerate(self.reg_param_list):
+                est_hs = HSTreeClassifier(base_est, reg_param)
+                est_hs.fit(X_in, y_in)
+                self.scores_[i].append(log_loss(y_out, est_hs.predict_proba(X_out)))
+        self.scores_ = [np.mean(s) for s in self.scores_]
+        self.reg_param = self.reg_param_list[np.argmin(self.scores_)]
```

Comment:
np.argmin won't work for custom scorers if they require argmax. ditto for regression